// tldraw/dist-cjs/lib/ui/components/primitives/TldrawUiTooltip.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TldrawUiTooltip_exports = {};
__export(TldrawUiTooltip_exports, {
  TldrawUiTooltip: () => TldrawUiTooltip,
  TldrawUiTooltipProvider: () => TldrawUiTooltipProvider,
  tooltipManager: () => tooltipManager
});
module.exports = __toCommonJS(TldrawUiTooltip_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_editor = require("@tldraw/editor");
var import_radix_ui = require("radix-ui");
var import_react = __toESM(require("react"), 1);
var import_layout = require("./layout");
const DEFAULT_TOOLTIP_DELAY_MS = 700;
class TooltipManager {
  static instance = null;
  currentTooltip = (0, import_editor.atom)("current tooltip", null);
  destroyTimeoutId = null;
  static getInstance() {
    if (!TooltipManager.instance) {
      TooltipManager.instance = new TooltipManager();
    }
    return TooltipManager.instance;
  }
  showTooltip(tooltipId, content, targetElement, side, sideOffset, showOnMobile, delayDuration) {
    if (this.destroyTimeoutId) {
      clearTimeout(this.destroyTimeoutId);
      this.destroyTimeoutId = null;
    }
    this.currentTooltip.set({
      id: tooltipId,
      content,
      side,
      sideOffset,
      showOnMobile,
      targetElement,
      delayDuration
    });
  }
  updateCurrentTooltip(tooltipId, update) {
    this.currentTooltip.update((tooltip) => {
      if (tooltip?.id === tooltipId) {
        return update(tooltip);
      }
      return tooltip;
    });
  }
  hideTooltip(editor, tooltipId, instant = false) {
    const hide = () => {
      if (this.currentTooltip.get()?.id === tooltipId) {
        this.currentTooltip.set(null);
        this.destroyTimeoutId = null;
      }
    };
    if (editor && !instant) {
      this.destroyTimeoutId = editor.timers.setTimeout(hide, 300);
    } else {
      hide();
    }
  }
  hideAllTooltips() {
    this.currentTooltip.set(null);
    this.destroyTimeoutId = null;
  }
  getCurrentTooltipData() {
    const currentTooltip = this.currentTooltip.get();
    if (!currentTooltip) return null;
    if (!this.supportsHover() && !currentTooltip.showOnMobile) return null;
    return currentTooltip;
  }
  supportsHoverAtom = null;
  supportsHover() {
    if (!this.supportsHoverAtom) {
      const mediaQuery = window.matchMedia("(hover: hover)");
      const supportsHover = (0, import_editor.atom)("has hover", mediaQuery.matches);
      this.supportsHoverAtom = supportsHover;
      mediaQuery.addEventListener("change", (e) => {
        supportsHover.set(e.matches);
      });
    }
    return this.supportsHoverAtom.get();
  }
}
const tooltipManager = TooltipManager.getInstance();
const TooltipSingletonContext = (0, import_react.createContext)(false);
function TldrawUiTooltipProvider({ children }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.Tooltip.Provider, { skipDelayDuration: 700, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(TooltipSingletonContext.Provider, { value: true, children: [
    children,
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TooltipSingleton, {})
  ] }) });
}
function TooltipSingleton() {
  const [isOpen, setIsOpen] = (0, import_react.useState)(false);
  const triggerRef = (0, import_react.useRef)(null);
  const isFirstShowRef = (0, import_react.useRef)(true);
  const editor = (0, import_editor.useMaybeEditor)();
  const currentTooltip = (0, import_editor.useValue)(
    "current tooltip",
    () => tooltipManager.getCurrentTooltipData(),
    []
  );
  const cameraState = (0, import_editor.useValue)("camera state", () => editor?.getCameraState(), [editor]);
  (0, import_react.useEffect)(() => {
    if (cameraState === "moving" && isOpen && currentTooltip) {
      tooltipManager.hideTooltip(editor, currentTooltip.id, true);
    }
  }, [cameraState, isOpen, currentTooltip, editor]);
  (0, import_react.useEffect)(() => {
    function handleKeyDown(event) {
      if (event.key === "Escape" && currentTooltip && isOpen) {
        tooltipManager.hideTooltip(editor, currentTooltip.id);
        event.stopPropagation();
      }
    }
    document.addEventListener("keydown", handleKeyDown, { capture: true });
    return () => {
      document.removeEventListener("keydown", handleKeyDown, { capture: true });
    };
  }, [editor, currentTooltip, isOpen]);
  (0, import_react.useEffect)(() => {
    let timer = null;
    if (currentTooltip && triggerRef.current) {
      const activeRect = currentTooltip.targetElement.getBoundingClientRect();
      const trigger = triggerRef.current;
      trigger.style.position = "fixed";
      trigger.style.left = `${activeRect.left}px`;
      trigger.style.top = `${activeRect.top}px`;
      trigger.style.width = `${activeRect.width}px`;
      trigger.style.height = `${activeRect.height}px`;
      trigger.style.pointerEvents = "none";
      trigger.style.zIndex = "9999";
      if (isFirstShowRef.current) {
        timer = setTimeout(() => {
          setIsOpen(true);
          isFirstShowRef.current = false;
        }, currentTooltip.delayDuration);
      } else {
        setIsOpen(true);
      }
    } else {
      setIsOpen(false);
      isFirstShowRef.current = true;
    }
    return () => {
      if (timer !== null) {
        clearTimeout(timer);
      }
    };
  }, [currentTooltip]);
  if (!currentTooltip) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_radix_ui.Tooltip.Root, { open: isOpen, delayDuration: 0, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.Tooltip.Trigger, { asChild: true, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { ref: triggerRef }) }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_radix_ui.Tooltip.Content,
      {
        className: "tlui-tooltip",
        side: currentTooltip.side,
        sideOffset: currentTooltip.sideOffset,
        avoidCollisions: true,
        collisionPadding: 8,
        dir: "ltr",
        children: [
          currentTooltip.content,
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.Tooltip.Arrow, { className: "tlui-tooltip__arrow" })
        ]
      }
    )
  ] });
}
const TldrawUiTooltip = (0, import_react.forwardRef)(
  ({
    children,
    content,
    side,
    sideOffset = 5,
    disabled = false,
    showOnMobile = false,
    delayDuration
  }, ref) => {
    const editor = (0, import_editor.useMaybeEditor)();
    const tooltipId = (0, import_react.useRef)((0, import_editor.uniqueId)());
    const hasProvider = (0, import_react.useContext)(TooltipSingletonContext);
    const enhancedA11yMode = (0, import_editor.useValue)(
      "enhancedA11yMode",
      () => editor?.user.getEnhancedA11yMode(),
      [editor]
    );
    const orientationCtx = (0, import_layout.useTldrawUiOrientation)();
    const sideToUse = side ?? orientationCtx.tooltipSide;
    (0, import_react.useEffect)(() => {
      const currentTooltipId = tooltipId.current;
      return () => {
        if (hasProvider) {
          tooltipManager.hideTooltip(editor, currentTooltipId, true);
        }
      };
    }, [editor, hasProvider]);
    (0, import_react.useLayoutEffect)(() => {
      if (hasProvider && tooltipManager.getCurrentTooltipData()?.id === tooltipId.current) {
        tooltipManager.updateCurrentTooltip(tooltipId.current, (tooltip) => ({
          ...tooltip,
          content,
          side: sideToUse,
          sideOffset,
          showOnMobile
        }));
      }
    }, [content, sideToUse, sideOffset, showOnMobile, hasProvider]);
    if (disabled || !content) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children });
    }
    let delayDurationToUse;
    if (enhancedA11yMode) {
      delayDurationToUse = 0;
    } else {
      delayDurationToUse = delayDuration ?? (editor?.options.tooltipDelayMs || DEFAULT_TOOLTIP_DELAY_MS);
    }
    if (!hasProvider || enhancedA11yMode) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_radix_ui.Tooltip.Root,
        {
          delayDuration: delayDurationToUse,
          disableHoverableContent: !enhancedA11yMode,
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.Tooltip.Trigger, { asChild: true, ref, children }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
              import_radix_ui.Tooltip.Content,
              {
                className: "tlui-tooltip",
                side: sideToUse,
                sideOffset,
                avoidCollisions: true,
                collisionPadding: 8,
                dir: "ltr",
                children: [
                  content,
                  /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.Tooltip.Arrow, { className: "tlui-tooltip__arrow" })
                ]
              }
            )
          ]
        }
      );
    }
    const child = import_react.default.Children.only(children);
    (0, import_editor.assert)(import_react.default.isValidElement(child), "TldrawUiTooltip children must be a single element");
    const handleMouseEnter = (event) => {
      child.props.onMouseEnter?.(event);
      tooltipManager.showTooltip(
        tooltipId.current,
        content,
        event.currentTarget,
        sideToUse,
        sideOffset,
        showOnMobile,
        delayDurationToUse
      );
    };
    const handleMouseLeave = (event) => {
      child.props.onMouseLeave?.(event);
      tooltipManager.hideTooltip(editor, tooltipId.current);
    };
    const handleFocus = (event) => {
      child.props.onFocus?.(event);
      tooltipManager.showTooltip(
        tooltipId.current,
        content,
        event.currentTarget,
        sideToUse,
        sideOffset,
        showOnMobile,
        delayDurationToUse
      );
    };
    const handleBlur = (event) => {
      child.props.onBlur?.(event);
      tooltipManager.hideTooltip(editor, tooltipId.current);
    };
    const childrenWithHandlers = import_react.default.cloneElement(children, {
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      onFocus: handleFocus,
      onBlur: handleBlur
    });
    return childrenWithHandlers;
  }
);
